autoendpoint/src/server/extractors/subscription.rs

```diff
@@ -1,66 +1,158 @@
-use crate::error::{ApiError, ApiErrorKind};
+use crate::error::{ApiError, ApiErrorKind, ApiResult};
 use crate::server::extractors::token_info::TokenInfo;
-use crate::server::ServerState;
+use crate::server::headers::crypto_key::CryptoKeyHeader;
+use crate::server::headers::vapid::{VapidHeader, VapidVersionData};
+use crate::server::{ServerState, VapidError};
 use actix_http::{Payload, PayloadStream};
 use actix_web::web::Data;
 use actix_web::{FromRequest, HttpRequest};
-use futures::future;
+use cadence::{Counted, StatsdClient};
+use futures::future::LocalBoxFuture;
+use futures::FutureExt;
+use openssl::hash;
+use std::borrow::Cow;
 
 /// Extracts subscription data from `TokenInfo` and verifies auth/crypto headers
 pub struct Subscription {
     pub uaid: String,
     pub channel_id: String,
-    pub api_version: String,
+    pub vapid: Option<VapidHeader>,
     pub public_key: Option<String>,
 }
 
 impl FromRequest for Subscription {
     type Error = ApiError;
-    type Future = future::Ready<Result<Self, Self::Error>>;
+    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
     type Config = ();
 
     fn from_request(req: &HttpRequest, _: &mut Payload<PayloadStream>) -> Self::Future {
-        // Collect token info and server state
-        let token_info = match TokenInfo::extract(req).into_inner() {
-            Ok(t) => t,
-            Err(e) => return future::err(e),
-        };
-        let state: Data<ServerState> = Data::extract(req)
-            .into_inner()
-            .expect("No server state found");
-        let fernet = state.fernet.as_ref();
-
-        // Decrypt the token
-        let token = match fernet.decrypt(&token_info.token) {
-            Ok(t) => t,
-            Err(_) => return future::err(ApiErrorKind::InvalidToken.into()),
-        };
-
-        if token_info.api_version == "v1" && token.len() != 32 {
-            // Corrupted token
-            return future::err(ApiErrorKind::InvalidToken.into());
-        }
+        let req = req.clone();
+
+        async move {
+            // Collect token info and server state
+            let token_info = TokenInfo::extract(&req).await?;
+            let state: Data<ServerState> =
+                Data::extract(&req).await.expect("No server state found");
+            let fernet = state.fernet.as_ref();
+
+            // Decrypt the token
+            let token = fernet
+                .decrypt(&repad_base64(&token_info.token))
+                .map_err(|_| ApiErrorKind::InvalidToken)?;
+
+            // Parse VAPID and extract public key
```

Comment:
nit: I think this block could be restricted to only running when `api_version == "v2"`.

Reply:
VAPID can still be provided for v1 endpoints, so if we gate VAPID on v2 we will lose metrics for VAPID usage on v1 endpoints. Is that expected?

Reply:
No, we want to know overall vapid use, not just how many use it for subscription locking. Likewise, we want to disallow invalid VAPID headers because it breaks the agreement around VAPID use (even if we're not using the data).

Reply:
> VAPID can still be provided for v1 endpoints, so if we gate VAPID on v2 we will lose metrics for VAPID usage on v1 endpoints (in addition to allowing invalid VAPID headers). Is that expected?

Ah I've misread the Python, you are correct